### ISSUE TITLE
(#feat): Information about the adviced method

### ISSUE THREAD

**Issue by konstantinkrassmann** [amoilanen/jsAspect#1]
Hi,
I just found your blogpost and made the library work in my JSFiddle: http://jsfiddle.net/B57MV/1/

The only problem i see there, that is missing to make this library use in a productive enviroment is, that i need more information about the adviced method and control to interrupt the process flow.

Something like its name. For a logging/tracing reason, you can't use this library yet.

Store and pass additional variables to the callback.

```
jsAspect.inject(Account, jsAspect.pointcuts.prototypeMethods, jsAspect.advices.before, 
  function(joinPointContext){
      console.log(joinPointContext.getFuncName()); // name of the adviced function
     console.log(joinPointContext.getFuncArguments()); // returns the arguments passed to the adviced function
      joinPointContext.stopExecution();  //stops the execution
  }
);
```

**Comment by amoilanen** [comment]
Hi,

Thank you for the feedback. I also plan to publish this library as a Node.js module, add a build that will produce a minified version usable in a browser, etc. For now it is still a work in progress but almost finished.

I like in the solution you suggested that we have access to the original method name and can expose some additional methods that will, for example, control the execution. Maybe I would also slightly change the names of some of the functions and even change them to fields like 'methodName', 'methodArguments'. Also 'stopExecution' can be just 'stop'. But this is a matter of taste.

The only thing I have doubts about is that now instead of transparently having all the arguments in the advice function we have to access them via `getFuncArguments()`. For example, instead of:

``` javascript
jsAspect.inject(Math, jsAspect.pointcuts.prototypeMethods, jsAspect.advices.before,
  function(x, y) {
    console.log("Performing operation ", x, y);
  }
);
```

we will have

``` javascript
jsAspect.inject(Math, jsAspect.pointcuts.prototypeMethods, jsAspect.advices.before,
  function(context) {
    var x = context.methodArguments[0];
    var y = context.methodArguments[1];

    console.log("Performing operation ", context.methodName, x, y);
  }
);
```

Handling special cases should not affect the basic API of the library much and should leave it simple in my opinion.

We may try to think of some solution that will take the best from both approaches. For example, it is possible to add context as an optional last argument that can be omitted but contains `methodName` and some useful methods like `stop`:

``` javascript
jsAspect.inject(Math, jsAspect.pointcuts.prototypeMethods, jsAspect.advices.before,
  function(x, y, context) {
    console.log("Stopping operation ", context.methodName, x, y);
    context.stop();
  }
);
```

What do you think? Please, let me know if you have some thoughts, I will try to implement this feature in the next few days.

**Comment by amoilanen** [comment]
In the revision 402466f7de336cf7315abedf6bd6689a6ea2f1eb the functionality was implemented for 'before' advices, I will also support the remaining 'after', 'around', 'afterThrowing', 'afterReturning' and then will close the issue if you think the implementation is acceptable.

**Comment by konstantinkrassmann** [comment]
Hi, 
Thank you for the reply. 

I will get the latest revision and try it out ;) 

I gonna port your framework as an angularJS module and make the 'after' work with promises, if it suits my needs now  ;)

You should improve the repositories documentation. Your implementation is the best i discovered yet.

**Comment by konstantinkrassmann** [comment]
Just checked your code:

http://jsfiddle.net/B57MV/3/

The problem is about last argument is the context: 
You place an beforeCallback before each method. The methods parameter amount can be variable. 
So you basically always have to do a workaround to access the context parameter:

``` javascript
jsAspect.inject(Account, jsAspect.pointcuts.prototypeMethods, jsAspect.advices.after,
    function afterCallback() {
      var context = arguments[arguments.length-1];
      //...
    }
);
```

That's not what you really want to do. 

This idea for a context comes from AspectJ for javascript. 

For me this is the best approach: 

``` javascript
jsAspect.inject(Math, jsAspect.pointcuts.prototypeMethods, jsAspect.advices.before,
  function(context) {
    var x = context.methodArguments[0];
    var y = context.methodArguments[1];

    console.log(context.className, " performs ", context.methodName, ". arguments:");
    console.log(context.methodArguments);
  }
);
```

**Or** make the context the first parameter, thats much better: 

``` javascript
jsAspect.inject(Math, jsAspect.pointcuts.prototypeMethods, jsAspect.advices.before,
  function(context, x, y) {
    console.log(context.className, " performs ", context.methodName, ". arguments:");
    console.log(context.methodArguments);
  }
);
```

Logs for my jsfiddle: 

`Account performs withDraw. Arguments: [1000]`

Here `className` is the `typeof` the first passed parameter of `inject`. (Check pull request #2 )

**Comment by amoilanen** [comment]
This makes, sense, I like the last option most:

``` javascript
jsAspect.inject(Math, jsAspect.pointcuts.prototypeMethods, jsAspect.advices.before,
  function(context, x, y) {
    console.log(context.className, " performs ", context.methodName, ". arguments:");
    console.log(context.methodArguments);
  }
);
```

This way we can still get the original arguments without being forced to call `context.methodArguments` while having direct access to the `context` argument. The implementation of `before` has been updated in the commit  707da6dd5f06651a98764d13d6ee0b8d40921404 Then I also re-factored the code a bit 2ffb7ee19c4f784add81844d50ffc3bb5ace5ff0

I will make similar changes for other advices in the next revisions.

**Comment by konstantinkrassmann** [comment]
That's perfect. 

Looking to have access in all join points.

One thing i noticed, that causes me to have a workaround all the time: 

``` javascript
 jsAspect.inject(Account, jsAspect.pointcuts.prototypeMethods, jsAspect.advices.before, 
  function(context, x){      
     console.log("calling ",context.methodName, " from ", context.targetConstructor.name, " with arguments");
     console.log(Array.prototype.slice.call(arguments,1));   //That's the only way to just logg all arguments
  }
);
```

[Example in jsfiddle](http://jsfiddle.net/B57MV/5/)

The amount of parameter is variable and i do place this advice code on all methods. 
It's just reducing a workaround but maybe give access to all parameter in the context and the function.

``` javascript
 jsAspect.inject(Account, jsAspect.pointcuts.prototypeMethods, jsAspect.advices.before, 
  function(context, x){      
     console.log("calling ",context.methodName, " from ", context.targetConstructor.name, " with arguments");
     console.log(context.methodArguments);   
  }
);
```

Maybe rename it to `context.method.name` and `context.method.arguments`

**Comment by amoilanen** [comment]
Makes sense. I will change it to `context.method.name`, `context.method.arguments`, `context.target.constructor.name`, etc.

Also probably we still have to provide a cross-browser solution for determining the name of the target constructor.

**Comment by konstantinkrassmann** [comment]
About the name of the constructor: 

I found this code: 

``` javascript
Object.prototype.getName = function() { 
   var funcNameRegex = /function (.{1,})\(/;
   var results = (funcNameRegex).exec((this).constructor.toString());
   return (results && results.length > 1) ? results[1] : "";
};
```

From the post here: [Stackoverflow](http://stackoverflow.com/questions/332422/how-do-i-get-the-name-of-an-objects-type-in-javascript)

Playing around in my [JSFiddle](http://jsfiddle.net/L7Fk8/) it seems to do exactly what we want. (Check dev-console F12)

Maybe make it possible to pass your own class name description: 

``` javascript
aspect.applyTo(someConstructor, "BasicContructor");
```

**Comment by amoilanen** [comment]
Implemented determining the function name in 3777920772b7bc343e3af3fc7d07a818c8c4deea and https://github.com/antivanov/jsAspect/commit/d98166c4270cac9fbda250d2bec7d238dd81b2a3
